### PR TITLE
New endpoint 'places' for looking up multiple osm_id's address #222

### PR DESCRIPTION
**Pull request by saerdnaer**

### CODE REVIEW

File: website/places.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+			echo 'Bulk User: Only ' .  CONST_ReverseSearch_Max_IDs . " ids are allowed in one request.\n";
+			exit;
+		}
+		
+		$type = ''; 
+		$id = 0;
+		$aPlaces = array();
+		foreach ($osm_ids AS $item) 
+		{
+			$type = $item[0];
```
Comment (lonvia) [inline website/places.php]: Fails for empty items (e.g. places.php?osmids=N1,,N2)

File: website/places.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	{
+		$oPlaceLookup = new PlaceLookup($oDB);
+		$oPlaceLookup->setLanguagePreference($aLangPrefOrder);
+		$oPlaceLookup->setIncludeAddressDetails($bShowAddressDetails);
+		
+		$osm_ids = explode(',', $_GET['osm_ids']);
+		
+		if ( count($osm_ids) > CONST_Places_Max_ID_count ) 
+		{
+			echo 'Bulk User: Only ' .  CONST_ReverseSearch_Max_IDs . " ids are allowed in one request.\n";
```
Comment (lonvia) [inline website/places.php]: Should that be CONST_Places_Max_ID_count?

File: lib/template/places-json.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
+<?php
+	$aFilteredPlaces = array();
+
+	if (!sizeof($aPlaces))
+	{
+		if (isset($sError))
+			$aFilteredPlaces['error'] = $sError;
+		else
+			$aFilteredPlaces['error'] = 'Unable to geocode';
```
Comment (lonvia) [inline lib/template/places-json.php]: I think it is better to return an empty list if none of the places could be found. (Same for XML response.) Is $sError even set anywhere?

File: website/places.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+	if (strpos(CONST_BulkUserIPs, ','.$_SERVER["REMOTE_ADDR"].',') !== false)
+	{
+		$fLoadAvg = getLoadAverage();
+		if ($fLoadAvg > 2) sleep(60);
+		if ($fLoadAvg > 4) sleep(120);
+		if ($fLoadAvg > 6)
+		{
+			echo "Bulk User: Temporary block due to high server load\n";
+			exit;
```
Comment (lonvia) [inline website/places.php]: Use userError() here, which creates a proper 400 response. Same below.

**Comment by lonvia** [comment]
I've added comments to the code for a few minor things. Other things I noticed:

 * I get odd results, when querying a nonexistent object, e.g. /places.php?osm_ids=N1. 
 * In the xml response the addressdetails are outside the result. The whole thing should look more like the answers to search queries, maybe even calling it `place` instead of `result`.

Could I also ask you to add basic tests similar to `tests/features/api/reverse_simple.php`? The minimum would be to test for valid json/xml. Bonus points for testing different input parameters.

File: website/places.php (hunk 1 above)
Comment (saerdnaer) [inline website/places.php]: In my tests it did work. I only get an "Notice: Uninitialized string offset" when I add error_reporting(E_ALL); 
But I will add some code to handle this case better.

File: website/places.php (hunk 2 above)
Comment (saerdnaer) [inline website/places.php]: of course

File: lib/template/places-json.php (hunk 3 above)
Comment (saerdnaer) [inline lib/template/places-json.php]: I basically used https://github.com/twain47/Nominatim/blob/master/lib/template/address-jsonv2.php as base for my new file. This part was copied from there.

**Comment by saerdnaer** [comment]
> "I get odd results, when querying a nonexistent object, e.g. /places.php?osm_ids=N1."
can you explain or give an more details example what happens?

I can't really test the code in production, as I don't have a good testing environment.

**Comment by mtmail** [comment]
There are some syntax errors in PHP (one bracket not closed) and Python (method called with too many variables). It's hard to write code when, as you say, you don't have a good testing environment setup. I will post my changes in a couple of hours and then it should be ready to be merged. All thanks goes to you of course!

@lonvia: the CONST_Places_Max_ID_count is set to 300 in settings.php. On my machine (with a remote worldwide database) looking for 300 existing nodes took 30 seconds. Should the limit be lower by default to avoid slowing down default Nominatim installations?

**Comment by lonvia** [comment]
Yeah, I think a limit of 50 is sufficient.

As for test environment, it would be completely sufficient to have a very small Nominatim installation of, say, Liechtenstein to code and test such a change. Of course, that still means going through the whole installation procedure, which isn't the most simple or fast thing to do. I wonder if it's time to have a look at docker to provide a convenient environment where you only have to run the setup script.

**Comment by mtmail** [comment]
I have https://www.vagrantup.com/ scripts for Nominatim I can share.

**Comment by saerdnaer** [comment]
This pull request can be closed now, as #283 was merged which implemented the same feature.

**Pull request closed without merging**